src/mod_ping.erl

```diff
@@ -150,132 +121,28 @@ terminate(_Reason, #state{host = Host}) ->
-    {noreply, State#state{timers = Timers}};
-handle_cast({stop_ping, JID}, State) ->
-    Timers = del_timer(JID, State#state.timers),
-    {noreply, State#state{timers = Timers}};
-handle_cast({iq_pong, JID, timeout}, State) ->
-    Timers = del_timer(JID, State#state.timers),
-    ejabberd_hooks:run(user_ping_timeout, State#state.host, [JID]),
-    case State#state.timeout_action of
-        kill ->
-            #jid{user = User, server = Server, resource = Resource} = JID,
-            case ejabberd_sm:get_session_pid(User, Server, Resource) of
-                Pid when is_pid(Pid) ->
-                    ejabberd_c2s:stop(Pid);
-                _ ->
-                    ok
-            end;
-        _ ->
-            ok
-    end,
-    {noreply, State#state{timers = Timers}};
-handle_cast(_Msg, State) ->
-    {noreply, State}.
-
-handle_info({timeout, _TRef, {ping, JID}},
-            #state{ping_req_timeout = PingReqTimeout} = State) ->
-    IQ = #iq{type = get,
-             sub_el = [#xmlel{name = <<"ping">>,
-                              attrs = [{<<"xmlns">>, ?NS_PING}]}]},
-    Pid = self(),
-    F = fun(_From, _To, Acc, Response) ->
-                gen_server:cast(Pid, {iq_pong, JID, Response}),
-                Acc
-        end,
-    From = jid:make(<<"">>, State#state.host, <<"">>),
-    Acc = mongoose_acc:new(#{ location => ?LOCATION,
-                              lserver => State#state.host,
-                              from_jid => From,
-                              to_jid => JID,
-                              element => jlib:iq_to_xml(IQ) }),
-    ejabberd_local:route_iq(From, JID, Acc, IQ, F, PingReqTimeout),
-    Timers = add_timer(JID, State#state.ping_interval, State#state.timers),
-    {noreply, State#state{timers = Timers}};
-handle_info(_Info, State) ->
-    {noreply, State}.
-
-code_change(_OldVsn, State, _Extra) ->
-    {ok, State}.
-
 %%====================================================================
 %% Hook callbacks
 %%====================================================================
-iq_ping(_From, _To, Acc, #iq{type = Type, sub_el = SubEl} = IQ) ->
-    case {Type, SubEl} of
-        {get, #xmlel{name = <<"ping">>}} ->
-            {Acc, IQ#iq{type = result, sub_el = []}};
-        _ ->
-            {Acc, IQ#iq{type = error, sub_el = [SubEl, mongoose_xmpp_errors:feature_not_implemented()]}}
-    end.
 
-user_online(Acc, _SID, JID, _Info) ->
-    start_ping(JID#jid.lserver, JID),
-    Acc.
+iq_ping(_From, _To, Acc, #iq{type = get, sub_el = #xmlel{name = <<"ping">>}} = IQ) ->
+    {Acc, IQ#iq{type = result, sub_el = []}};
+iq_ping(_From, _To, Acc, #iq{sub_el = SubEl} = IQ) ->
+    {Acc, IQ#iq{type = error, sub_el = [SubEl, mongoose_xmpp_errors:feature_not_implemented()]}}.
 
-user_offline(Acc, _SID, JID, _Info, _Reason) ->
-    stop_ping(JID#jid.lserver, JID),
+user_online(Acc, {_, Pid} = _SID, _JID, _Info) ->
+    ejabberd_c2s_info_handler:add(Pid, mod_ping, ?MODULE, handle_info, undefined),
```

Comment:
@fenek how about we start the timer when adding the handler? In this case we avoid the `undefined` situation and timer is started anyway when we call the `init` function of the handler.